Restore changes to parse arrays of images

denisgoryaynov opened the pull request:
It seems that the changes form my previous pull request were overwritten by this [commit](https://github.com/strapi/gatsby-source-strapi/commit/6c313ebf81fcf6899ea4658c306c87a4421394b4) that implemented compatibility with Gatsby Cloud. This pull request is just to restore the files that allowed parsing arrays of images and query them like this

```
{
  allStrapiArticle {
    edges {
      node {
        id
        singleImage {
         publicURL
        }
        multipleImages {
          localFile {
            publicURL
          }
        }
      }
    }
  }
}
```

alexandrebodin commented:
This wasn't overriden but moved higher in the file. I don't get what your PR is supposed to add

denisgoryaynov commented:
@alexandrebodin hi, this PR allows to use the publicURL field on arrays of images. Right now if do something like this:
```
query Test {
 allStrapiProject {
    nodes {
      title
      multipleMedia {
        localFile {
          publicURL
        }
      }
      section {
        media {
          localFile {
            publicURL
          }
        }
      }
    }
  }
}
```

gatsby fails to build because there are no nodes for the images. What this PR does is the same that my last PR did, #90 it allows to query the localFile field when using the Multiple Media type in the strapi admin panel.

alexandrebodin commented:
Ok, Thank you. I'll check this. The diff is kinda hard to read so will have to test this thouroughly :)

alexandrebodin commented:
@denisgoryaynov Can you fix the conflicts in yout PR ?

denisgoryaynov commented:
@alexandrebodin yes, I can do it later today

dsfrederic commented:
@denisgoryaynov is it possible to use Gatsby-image in combination with multiple files?

carlaiau commented:
@denisgoryaynov how is the conflict resolve going? Keep up the great work!

denisgoryaynov commented:
@alexandrebodin I fixed the conflicts, sorry if it took me so long

alexandrebodin commented on src/normalize.js:
```diff
@@ -5,68 +5,97 @@ const extractFields = async (
-        }
+    // If we have cached media data and it wasn't modified, reuse
+    // previously created file node to not try to redownload
+    if (cacheMediaData && item.updatedAt === cacheMediaData.updatedAt) {
```
Actually, this should be an option as the timestamps are configurable per model. (e.g by default on mongo it is camelcased `createdAt / updatedAt` and on sql databases it would be `created_at / updated_at`.)

denisgoryaynov replied:
Oh I didn't know that, I just changed it because at some point in the code the variables was set as updatedAt but then inside that if the value being used was updated_at which was always undefined for and I just though that it was an inconsistency. I can try to test it and fix it tomorrow.

denisgoryaynov replied:
@alexandrebodin I fixed it with your suggestion but I have a question.
What I did is to create a new variable like this: `const itemUpdatedAt = item.updatedAt || item.updated_at`, is there a better way to check what database is being used or this method should be fine?

alexandrebodin replied:
Hi, This should be fine for now. We will need to fetch model metadata to be more dynamic in the future though :) Will test this weekend  se we can release next week ;)

alexandrebodin commented:
HI @denisgoryaynov can you just fix the DCO check on this PR plz ? :D

denisgoryaynov commented:
@alexandrebodin I fixed all the issues. I used the localFile key because at the time I didn't know how to do it and that's what other plugins were using (for example gatsby-source-wordpress) and unfortunately I didn't really have much time to take a better look at it and see if it was possible to solve the issue without using the localFile key.

carlaiau commented:
Got ya branch working for a current project I'm working on using gatsby-image. Hope it gets pulled in, thanks for the great work!

peterfortuin commented:
I have also check out this branch in my current project and it solves my problem with arrays of images. Thanks for the great work.

peterfortuin commented:
@denisgoryaynov Is there a way to use your repository/branch into my Gatsby project with yarn instead of the standard gatsby-source-strapi package that is downloaded from npm?

carlaiau commented:
You can just yarn add the git directly but you need to fork it to your own and build the package. This is the src for the plugin so does nothing have the compiled JS that Gatsby is expecting. I forked the git, built the src and committed that and it works fine

Sent from ProtonMail Mobile

On Mon, Apr 27, 2020 at 4:23 AM, Peter <notifications@github.com> wrote:

> [@denisgoryaynov](https://github.com/denisgoryaynov) Is there a way to use your repository/branch into my Gatsby project with yarn instead of the standard gatsby-source-strapi package that is downloaded from npm?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, [view it on GitHub](https://github.com/strapi/gatsby-source-strapi/pull/118#issuecomment-619579354), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ABUTYF2SKWFWHOAJ7JP2SITRORNX7ANCNFSM4L24ENRA).

peterfortuin commented:
Thanks a lot @carlaiau . This would allow me to continue with my project.

denisgoryaynov commented:
@alexandrebodin that's a good idea, I recently updated to strapi beta 20  and didn't notice that these fields were missing from single images, I'll take a look a it!

denisgoryaynov commented:
@alexandrebodin Hi, I finally took some to fix the issue. In order to support extra image fields such as alternative I had to use the localFile key also on single images so instead of querying images like this: 
```
query {
   strapiArticle {
      single_image {
         publicURL
      }
   }
}
```
to access extra image fields the new query would be:
```
query {
   strapiArticle {
      single_image {
         alternativeText
         localFile {
            publicURL
         }
      }
   }
}
```

So basically the schema between single media type and multiple media type is the same now, but it's a breaking change. What do you think?

alexandrebodin commented:
Thank your for your hard work ! I will need to fiddle a bit with this before We can merge or release :) 

We can go to v1 if this makes sense no worries on that

denisgoryaynov commented:
@alexandrebodin Hi! Since I'm working a lot with strapi and gatsby I noticed that gatsby-source-strapi was using an old version of gatsby-source-filesystem so i moved it to peerDependencies and updated the code to support >2.x. I had to do this since gatsby-source-filesystem v1 was generating images and documents with a uid instead of a name and it was bad for the user to download a file called abc123.pdf; Updating gatsby-source-filesystem to v2 allows to give all files a name so the files are now called filename-hash.pdf

alexandrebodin commented:
@denisgoryaynov thanks for the update, can I review and test or you still want to add stuff to it ?

denisgoryaynov commented:
@alexandrebodin You can do it, I think it should be good now

tonoli commented:
@denisgoryaynov  Thank you for your work!
@alexandrebodin is this planed to get merged anytime soon?

ordishc commented:
@denisgoryaynov Thanks for looking into this!
@alexandrebodin has a fix been pushed into v3.0.5? Or does this still need to be merged?

alexandrebodin commented:
Hi, this repo si not in sync with the strapi version. I will be merging this work soon but there is some conflicts after I added eslint and prettier to make the project easier to maintain. if @denisgoryaynov is willing to update this would help !

denisgoryaynov commented:
@alexandrebodin Hi! I can take a look later today or tomorrow morning

ordishc commented:
> @alexandrebodin Hi! I can take a look later today or tomorrow morning

Did you manage to take a look in the end? No worries if not, I'm happy to look for a work around for the time being.

denisgoryaynov commented:
@ordishc Sorry, I ended up being full of work and things to study for university, I will take a look right now

denisgoryaynov commented:
@alexandrebodin I updated my code, it should be good now

ordishc commented:
Thanks @denisgoryaynov ! Much appreciated :)

tonoli commented on package.json:
```diff
@@ -55,6 +54,12 @@
+  "peerDependencies": {
+    "gatsby-source-filesystem": "^2.x"
+  },
+  "pre-commit": [
```
@denisgoryaynov

denisgoryaynov replied:
@alexandrebodin @tonoli Sorry for the delay, I was on vacation. I pushed the updated code now

hashinteractive commented:
Hey guys,

Any update on this PR as we are still only able to access the `thumbnail` format for multiple images as a local file in Gatsby. Such as:
![image](https://user-images.githubusercontent.com/12477388/92747993-53f64780-f34a-11ea-9606-b6d4411f3f66.png)

As you can see there are no local file options for multiple image formats only the `thumbnail` format. Also, the docs seem to be off:
```
{
  allStrapiArticle {
    edges {
      node {
        id
        singleImage {
         publicURL
        }
        multipleImages {
          localFile {
            publicURL
          }
        }
      }
    }
  }
}
```
As there is no query field `localFile` for multiple images.

denisgoryaynov commented:
@alexandrebodin @tonoli Hi, I fixed the conflict in fetch.js after the latest commit

hashinteractive commented:
@denisgoryaynov thanks for your work! I have not tested your PR but does it address creating a local gatsby File for each of the strap image formats (large, medium and thumbnail) or does it create a whole new field such as 
```
node {
  multipleImages {
    localFile
  }
}
```

Just curious how your PR implements the fix.

denisgoryaynov commented:
@hashinteractive Hi, yes it does create a localFile field for every format

<img width="1578" alt="Schermata 2020-09-11 alle 20 59 47" src="https://user-images.githubusercontent.com/16140813/92962754-be55d780-f471-11ea-9abe-2bdaed583336.png">

What the code does is to iterate every field recursively and if it finds an image (I test for the "mime" field to detect an image as the original implementation) it generates a localFile node with publicURL and childImageSharp properties

P.S. Thank you, when doing this screenshot I found out that there was a bug when generating images for formats.
My code and the original implementation used this to generate a cache key for every image:
`const mediaDataCacheKey = 'strapi-media-${item.id}'`
but images inside formats do not have an id field so every key was assigned a value of: strapi-media-undefined, and showed wrong images when generating a localFile node.
Using the hash field fixes the issue:
`const mediaDataCacheKey = 'strapi-media-${item.hash}'`

hashinteractive commented:
@denisgoryaynov thanks for your work and for the explanation. Hopefully we can get the PR merged soon 🙏

MrChrisRodriguez commented:
@alexandrebodin what's the timeline for getting this PR merged?

MrChrisRodriguez commented:
@denisgoryaynov I pulled your fork and when I build it fails because AllStrapiWebsites isn't showing up. When I revert back to origin AllStrapiWebsites appears. Have you tested your latest changes on a vanilla Gatsby-Strapi environment?

Bubiec commented:
Will this pull request fix fetching alt text from single images? I would really love to see that merged soon.

denisgoryaynov commented:
@MrChrisRodriguez Hi, sorry for the late response but I missed the notification. I tried by setting up a vanilla project using `gatsby new frontend` and `npx create-strapi-app admin` and as far as I can tell the build works correctly. Here are the nodes that are generated:
<img width="1600" alt="Schermata 2020-10-24 alle 15 58 06" src="https://user-images.githubusercontent.com/16140813/97083673-fd9b4a80-1611-11eb-8997-27477fb13a3e.png">

And here is what I added to my package.json:
`"gatsby-source-strapi": "git+https://github.com/denisgoryaynov/gatsby-source-strapi.git#build",`

I forgot to update to build branch with the latest files from the master branch so that's maybe why you had issues, it should be ok now.

@Bubiec yes it does generate alternativeText for single images, you can see it from the screenshot above

alexandrebodin commented:
Hi @denisgoryaynov. I refactored quite a lot of code and integrated changes to support what you have done.
Your code was a huge Help ! thanks a lot.

I released v1.0.0-alpha.0 with the latest changes. please try it out to validate things work for you so we can move release as latest !

denisgoryaynov commented:
Hi @alexandrebodin, thank you! I will try the new version as soon as I can!

DimaMiro commented:
Hello 👋 How is it going with this PR? @denisgoryaynov 
I do need this change in my project 🙂

alexandrebodin commented:
Hi @DimaMiro the changes of this PR were integrated in the v1 alpha release `v1.0.0-alpha.0`. Please try it out and give some feedback so we can move it to stable :)

DimaMiro commented:
> Hi @DimaMiro the changes of this PR were integrated in the v1 alpha release `v1.0.0-alpha.0`. Please try it out and give some feedback so we can move it to stable :)

Hi Alexandre 👋 I've tried this change in alpha and it works correctly. Thank you 🙌

strapi-cla commented:
[![CLA assistant check](https://cla.strapi.io/pull/badge/not_signed)](https://cla.strapi.io/strapi/gatsby-source-strapi?pullRequest=118) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.strapi.io/strapi/gatsby-source-strapi?pullRequest=118) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: denisgoryaynov<br/>:x: alexandrebodin<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.strapi.io/check/strapi/gatsby-source-strapi?pullRequest=118) it.</sub>

visuvelu commented:
Hi, how is it going with PR? @denisgoryaynov  @alexandrebodin I need this change in my projects.

alexandrebodin commented:
@visuvelu if you check the comments above you will see we made the fix in the alpha version but not many people have tester it so far so I'm holding off the release. You can use it as it fixes a lot of the current issues

Arahis commented:
Worked great for me. Thanks a lot!

alexandrebodin commented:
Hi, We just released the v1.0.0 fixing including this requested changes :)
